src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java

```diff
@@ -0,0 +1,283 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.shell.Subprocess;
+import com.google.devtools.build.lib.shell.SubprocessBuilder;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.File;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+/**
+ * An intermediate worker that receives response from the worker processes
+ */
+public class WorkerMultiplexer extends Thread {
+  private static Semaphore semInstanceMap = new Semaphore(1);
+  private static Map<Integer, WorkerMultiplexer> instanceMap = new HashMap<>();
+  private Map<Integer, InputStream> responseMap;
+  private Map<Integer, Semaphore> responseChecker;
+  private Semaphore semResponseMap;
+  private Semaphore semResponseChecker;
+  private Semaphore semAccessProcess;
+
+  private Subprocess process;
+  private Integer workerHash;
+
+  private Thread shutdownHook;
+
+  WorkerMultiplexer(Integer workerHash) {
+    semAccessProcess = new Semaphore(1);
+    semResponseMap = new Semaphore(1);
+    semResponseChecker = new Semaphore(1);
+    responseChecker = new HashMap<>();
+    responseMap = new HashMap<>();
+    this.workerHash = workerHash;
+
+    final WorkerMultiplexer self = this;
+    this.shutdownHook =
+      new Thread(
+        () -> {
+          try {
+            self.shutdownHook = null;
+            self.destroyMultiplexer();
+          } finally {
+            // We can't do anything here.
```

Comment:
OK you cannot. But what things would you expect to happen here, and why is it OK to ignore them?

And actually, you could log these with logger.warning() or something.

Reply:
FWIW, this appears to be copy-paste from https://github.com/bazelbuild/bazel/blob/9a32b861799278217c37eb32f864a951ca99617e/src/main/java/com/google/devtools/build/lib/worker/Worker.java#L59-L68

Reply:
This part has been removed.